src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb

```diff
@@ -3429,61 +3429,174 @@ class C
+                              </Document>, {provider})
 
                 state.Workspace.Options = state.Workspace.Options.WithChangedOption(
                     CompletionOptions.BlockForCompletionItems, LanguageNames.CSharp, False)
 
                 state.SendTypeChars("Sys")
-                state.SendCommitUniqueCompletionListItem()
-                Await Task.Delay(250)
-                Await state.AssertNoCompletionSession(block:=False)
-                Assert.Contains("Sys", state.GetLineTextFromCaretPosition())
-                Assert.DoesNotContain("System", state.GetLineTextFromCaretPosition())
 
-                tcs.SetResult(True)
+                If completionImplementation = CompletionImplementation.Legacy Then
+                    state.SendCommitUniqueCompletionListItem()
+                    Await Task.Delay(250)
 
-                Await state.WaitForAsynchronousOperationsAsync()
-                Await state.AssertNoCompletionSession()
-                Assert.Contains("System", state.GetLineTextFromCaretPosition())
+                    state.AssertNoCompletionSessionWithNoBlock()
+                    Assert.Contains("Sys", state.GetLineTextFromCaretPosition())
+                    Assert.DoesNotContain("System", state.GetLineTextFromCaretPosition())
+                    tcs.SetResult(True)
+
+                    Await state.WaitForAsynchronousOperationsAsync()
+                    Await state.AssertNoCompletionSession()
+                    Assert.Contains("System", state.GetLineTextFromCaretPosition())
+                Else
+                    Dim task1 As Task = Nothing
+                    Dim task2 As Task = Nothing
+
+                    Dim providerCalledHandler = Sub()
+                                                    task1 = Task.Run(Sub()
+                                                                         ' 2. Hang here as well.
+                                                                         Dim completionItem = state.GetSelectedItemOpt()
+                                                                     End Sub)
+                                                    task2 = Task.Run(Sub()
+                                                                         Thread.Sleep(250)
+                                                                         Try
+                                                                             ' 3. Check that the other task is running/hanging.
+                                                                             Assert.Equal(TaskStatus.Running, task1.Status)
+                                                                             Assert.Contains("Sys", state.GetLineTextFromCaretPosition())
+                                                                             Assert.DoesNotContain("System", state.GetLineTextFromCaretPosition())
+                                                                         Finally
```

Comment:
To avoid hangs waiting for the task to complete if any of asserts failed. Added a comment

---
In reply to: [246999978](https://github.com/dotnet/roslyn/pull/31534#discussion_r246999978) [](ancestors = 246999978)